feat!: merge quil-py into quil-rs

asaites opened the pull request:
Merges the `quil-py` crate into `quil-rs`:

- Removes the original `quil-py` crate entirely.
- Updates to the latest release of `pyo3`.
- Adds `pyo3` attributes directly to `quil-rs` items. 
- Removes the dependency on `rigetti-pyo3`, as its features are mostly subsumed by later releases of `pyo3`.
- Introduces a `python` feature that gates `pyo3`-related additions.
- Automatically generates Python type stubs for `quil`.
- Supports `copy`/`pickle` using `__getnewargs__` rather than `__setstate__`.
- Introduces a linting script to catch mistakes related to `pyo3` attributes within `quil-rs`.
- Updates documentation, READMEs, and CONTRIBUTING guide, and migration notes.
- Updates and simplifies CI workflows.
- Adds `cargo hack` to CI to check features compatibility.
- Adds script to check semver-incompatibile changes in `quil` API, and integrates it with CI.

Closes: #456
Closes: #461 
Closes: #463 
Closes: #148

Shadow53 commented on quil-rs/src/expression/mod.rs:
```diff
@@ -76,12 +120,15 @@ pub enum EvaluationError {
     Number(Complex64),
-    PiConstant,
+    #[pyo3(name = "pi")]
+    PiConstant(),
```
Why does this need empty parentheses? (I'm guessing it's something related to PyO3, but would be nice to have a comment here)

Shadow53 commented on quil-rs/src/expression/mod.rs:
```diff
@@ -145,20 +207,27 @@ impl InfixExpression {
     }
 }
 
+// TODO: PartialEq/Eq is inconsistent with Hash: https://github.com/rigetti/quil-rs/issues/458
```
😬

asaites replied:
Yeah, the good news is, it _probably_ doesn't come up very often.

antalsz replied:
```suggestion
// TODO (#458): PartialEq/Eq is inconsistent with Hash
```

I think this tends to be how we record things like this in code – Kalan once advised me to avoid unannotated TODOs

Shadow53 commented on quil-rs/src/expression/mod.rs:
```diff
@@ -764,7 +901,7 @@ pub mod interned {
     use super::*;
 
     macro_rules! atoms {
-        ($($func:ident: $ctor:ident$(($typ:ty))?),+ $(,)?) => {
+        ($($func:ident: $ctor:ident($($typ:ty)?)),+ $(,)?) => {
```
What motivated this change?

asaites replied:
The addition of the empty tuple variants -- now this needs to handle `()` with no arguments.

Shadow53 commented on quil-rs/src/instruction/classical.rs:
```diff
@@ -118,6 +128,8 @@ impl Quil for BinaryOperand {
 
 #[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
+#[pyclass(module = "quil.instructions", eq, frozen, hash, get_all)]
+#[pyo3(rename_all = "snake_case")]
```
Shouldn't this be `"SCREAMING_SNAKE_CASE"`? I believe that's the Python convention

Shadow53 commented on quil-rs/src/instruction/classical.rs:
```diff
@@ -99,6 +107,8 @@ impl Quil for ArithmeticOperator {
 
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
+#[pyclass(module = "quil.instructions", eq, frozen, hash, get_all)]
+#[pyo3(rename_all = "snake_case")]
 pub enum BinaryOperand {
     LiteralInteger(i64),
```
Should this be `"SCREAMING_SNAKE_CASE"`? I'm less certain than the other comment on this because these variants have associated data

Shadow53 commented on quil-rs/src/instruction/mod.rs:
```diff
@@ -71,13 +76,239 @@ pub use self::reset::Reset;
+    Ok(())
+}
+
+impl_repr!(Arithmetic);
```
Why are these here and not next to the type definitions

asaites replied:
I was doing weird shell incantations to make sure I had correctly brought over all the `impl`s, so seeing them all in one place (and sorting them) made this easier for me to verify.

That said, I think I'd rather just have a `macro_rule!` for something like `impl_instruction!` that took a list of instruction types and expanded to the `impl_*`s that are needed, and maybe even the `add_class`es -- essentially a more specific version of the submodule implementation macro from `rigetti-pyo3`. I think that would be more readable, and much less error-prone.

Shadow53 commented on quil-rs/src/program/mod.rs:
```diff
@@ -86,69 +155,116 @@ type Result<T> = std::result::Result<T, ProgramError>;
         }
     }
 
+    /// Return a deep copy of the `Program`.
+    fn copy(&self) -> Self {
+        self.clone()
+    }
```
How is this different from deepcopy?

asaites replied:
I copied this [from quil-py](https://github.com/rigetti/quil-rs/blob/7f256085259c3b37160b73e67ff50e72523d7e40/quil-py/src/program/mod.rs#L89), but I agree that using `deepcopy` should (usually) give the same result.

That said, my gut is this is probably faster. Using `deepcopy` on a `Program` will end up routing through `pickle`, and the `__getstate__`/`__setstate__` implementations here require round-tripping through byte serialization.

One place these do differ: you can't `pickle` a program that has unresolved `QubitPlaceholder`s, but this `clone` should work just fine... and notice that it won't result in distinct placeholders, either, which I think is another point in the "we should probably get rid of that logic from `Instruction`s" basket.

As far as Pythonic-ness, for what it's worth, `dict`s have a `copy` method, too ... to produce [a shallow copy](https://docs.python.org/3/library/stdtypes.html#dict.copy)!

antalsz replied:
IMO, that's an argument for either removing `Program::copy` or making it implement a shallow copy; `Program` has mutating methods, so a shallow copy is meaningful there.

Shadow53 replied:
Not going to go into what kinds of copying we _should_ have, but where we enable copying we should follow the convention of (what seems to be) `copy == shallow`, `deepcopy == full clone`

This may cause a little bit of complexity/verbosity for types containing `Arc`

Shadow53 commented on quil-rs/src/program/mod.rs:
```diff
@@ -86,69 +155,116 @@ type Result<T> = std::result::Result<T, ProgramError>;
+    }
+
+    #[getter]
+    // TODO: (9b9690d4 Marquess Valdez 2023-06-01 14:10:25 -0700) Should this filtering move to Program? Should we assume memory_regions will always make up all
```
Wonder if there's an issue for this

asaites replied:
I've now created one to link it to.

Shadow53 commented on quil-rs/src/waveform/templates.rs:
```diff
@@ -104,8 +167,44 @@ impl WaveformTemplate for ErfSquare {
 
+#[pymethods]
+impl ErfSquare {
+    #[new]
+    #[allow(clippy::too_many_arguments)]
+    fn __new__(
```
I don't remember the annotation for this, but there's a way to specify that all of these are keyword arguments, and I think that should be added to large constructors like this.

antalsz replied:
Unfortunately the only way I know to do that is

```
#[pyo3(signature = (*, duration, risetime, sample_rate, pad_left, pad_right, positive_polarity, scale, phase, detuning))]
```

I.e., respecifying every single argument name.  (And it's worse if you want to override the signature displayed from `help`, too.)

Docs: https://pyo3.rs/main/function/signature

asaites replied:
We definitely can modify it. Do you think they should _only_ support `kwargs`, and not positional at all?

Shadow53 replied:
For that many arguments, my preference would be to _only_ support `kwargs`

Shadow53 reviewed (commented):
Aside from the above comments, one note:

I was thinking we would end up needing something like `#[cfg(feature = "pyo3")]` everywhere to enable or disable PyO3 support. That is, pure-Rust users should not need PyO3 in their dependency tree.

asaites commented on quil-rs/Cargo.toml:
```diff
@@ -26,9 +29,14 @@ strum.workspace = true
+numpy = "0.25.0"
+pyo3 = { version = "0.25.1", features = ["indexmap", "num-complex", "multiple-pymethods"] }
+num-complex = "0.4.6"
+internment = { version = "0.8.6", features = ["arc"] }
```
I'm not sure exactly what the original comment was trying to get at, but didn't see a reason to drop it during the merge, so I know I could use some more context myself. Do you know of something specific here you'd like to see -- like a different comment or a link to particular docs?

antalsz replied:
As the person who added `internment` as a dependency: I suspect that what's happening is that because I used `cargo add`, it just chucked that package at the end of the section without regard for comments.  The same may be true of `num-complex`, since I don't think that's for PyO3.

asaites commented on quil-rs/pyproject.toml:
```diff
@@ -1,7 +1,7 @@
 [project]
 name = "quil"
-version = "0.16.0"
```
I'll keep that in mind, and remove the change before final merge. For the moment, it gives me confidence I've got the right one installed in the `venv`s I'm using to test elsewhere 😅

asaites replied:
Since this file moved from the old crate, I updated the `knope.toml` so the Rust crate and Python package are now in the same collection of "versioned_files". A consequence is that they'll need to have the same version number. I think this is probably a good thing (though it means a big version jump for the `quil` Python package), but let me know if you think there's a more appropriate solution.

Shadow53 commented:
Regarding making the `pyo3` dependency optional, here are the general changes I'd suggest. Some may seem obvious, but I'm listing everything just in case.

- Make `pyo3` and related dependencies `optional = true`
- Add a `pyo3` (or `python`) feature as `pyo3 = ["dep:pyo3"]`, adding any other dependencies
- Replace instances of `#[pyclass]` with `#[cfg_attr(feature = "pyo3", pyclass)]`
  - Similar for PyO3 `derive`, any `#[pyo3(...)]` attributes, etc.
- Any `impl` blocks that are the _same_ for both Rust and PyO3 should replace `#[pymethods]` with `#[cfg_attr(feature = "pyo3", pymethods)]`
- Any `impl` blocks that are _different_ for native Rust and PyO3 should leave the Rust-native block in place and add either:
  - An `impl` block with `#[cfg(feature = "pyo3")]` just below it
  - Or, if the file is getting too big, a separate submodule -- e.g. `python.rs` -- that contains all of the PyO3-only code. The module itself can be annotated `#[cfg(feature = "pyo3")]` instead

------

Actually, let me call that out explicitly: since we effectively have two versions of the library living in the same crate, I think anything that is specific to _only_ PyO3 should generally be placed in its own private submodule. Anything that should be public can be reexported in the parent module, if that makes sense.

That is, something like this:

```rust
// src/foo/mod.rs
#[cfg(feature = "pyo3")]
mod python; // Not `pyo3` as that would conflict with the crate `pyo3`

#[cfg_attr(feature = "pyo3", pyo3::pyclass)]
pub struct Foo {
    bar: String,
}

#[cfg_attr(feature = "pyo3", pyo3::pymethods)]
impl Foo {
    #[cfg_attr(feature = "pyo3", pyo3(new))]
    pub fn new(bar: String) -> Self {
        Self { bar }
    }
]

//  src/foo/python.rs
use pyo3::prelude::*;
use super::Foo;

#[pymethods]
impl Foo {
    pub fn python_only_bar(&self) -> &str {
        &self.bar
    }
}
```

---------

To make sure that everything works correctly with and without `pyo3` enabled, you can use [cargo hack](https://github.com/taiki-e/cargo-hack), which allows you to run `cargo check`, `cargo clippy, `cargo test`, etc. over multiple sets of features.

Shadow53 commented on quil-rs/examples/generate_test_expressions.rs:
```diff
@@ -68,7 +68,7 @@ fn parenthesized(expression: &Expression) -> String {
         ),
         Number(_) => format!("({})", expression.to_quil_or_debug()),
-        PiConstant => "pi".to_string(),
+        PiConstant() => "pi".to_string(),
```
Idle thought: would it be worth applying `#[cfg(...)]` to this enum variant so that the Rust code follows convention as `PiConstant` and the PyO3 code is `PiConstant()` to support the `pyclass` macro?

I'm not sure that it _would_ be worth, but it would be a way to keep the Rust side conventional

jselig-rigetti reviewed (commented):
I think you've nailed a good balance of being unobtrusive to the rust code while still enabling flexibility in both rust/python targets. The stub gen is also really awesome to see.

I see the `scripts/findpyclasses.py` aims to make it easier to make sure that `modules = "..."` strings are correct, so that's good - especially if it can be hooked up to pre-merge actions.

The passing of the `tests_py` test will give me confidence that we haven't introduced a regression. However, since some of the usages changes from like function calls to getters, I think this should probably be a breaking change (`feat!: ...`)

asaites commented on optipy/src/lib.rs:
```diff
@@ -0,0 +1,223 @@
+    let mut stripper = StripPyO3 { only_stubs: false };
+
+    let opt_parser = syn::meta::parser(|meta| {
+        if meta.path.is_ident("only_stubs") {
```
I didn't know, but I checked:

```
error: expected `,`
  --> quil-rs/src/expression/quilpy.rs:42:66
   |
42 | ...ip_pyo3(only_stubs = false))]
   |
```

asaites commented on optipy/src/lib.rs:
```diff
@@ -0,0 +1,223 @@
+                        || id == "pymethods"
+                        || id == "pyfunction"
+                        || id == "pymodule"
+                        || id == "gen_stub")
```
Likely it should; it just happens we're not using them.

There is a part of me that wonders if the overall approach would be better if more "generic", i.e., if instead of being specific to `pyo3` and `pyo3_stub_gen`, you could define the set of tags to remove dynamically using appropriate `#[cfg_attr(...)]`s, and then just have a `#[strip]` attribute on relevant items to process.

I can think of some ways of making that work, but I suspect it'd end up with more noise than it already has.

asaites commented on quil-rs/python/quil/expression.pyi:
```diff
@@ -0,0 +1,210 @@
+    class Address(Expression):
+        __match_args__ = ("_0",)
+        @property
+        def _0(self) -> MemoryReference: ...
```
I need to add some more clear documentation about the Python usage. The short of it is that in typical usage, you'd not access this via `_0` or `[0]` at all, but instead use the `match` syntax:

```python
match expr:
    case Expression.Address(mref):
        # do something with mref
    case FunctionCall(fexpr):
        ...
    case _:
        ...
 ```
 
 You could also do something like `isinstance(expr, Expression.Number)` and then access `expr._0`, knowing it's `complex`.

It's honestly a little annoying for `Instruction`s, though, because e.g. `assert not issubclass(instructions.Gate, Instruction.Gate)` and vice versa -- `Instruction.Gate` is a subclass of `Instruction` which has a single attribute, `_0` which is of type `instructions.Gate`. And so constructing it feels weird:

```python
gate = Gate("X", (), (Qubit.Fixed(0),), ())
instr = Instruction.Gate(gate)

assert isinstance(gate, Gate)
assert isinstance(instr, Instruction)
assert isinstance(instr, Instruction.Gate)
assert not isinstance(instr, Gate)
assert not isinstance(gate, Instruction.Gate)
assert instr._0 == gate
```

The `_0` is the name that `pyo3` generates automatically. If we used `struct` variants instead, we could name the fields, but it'd break a lot more code.

The other option is to just let `Instruction` live exclusively in Rust-land, and have a special Python `BaseInstruction` `#[pyclass]` from which are subclassed the various instruction types. I don't know exactly what that would look like in the end, to handle the translation, but it may be worth it in the end to get more ergonomic and performant Python support. Even though it would mean less shared code, it'd still be benefit from having these packages united.

asaites commented on quil-rs/src/expression/quilpy.rs:
```diff
@@ -0,0 +1,222 @@
+
+    #[getter]
+    fn left(&self) -> Expression {
+        (*self.left).clone()
```
`Expression` isn't `Copy`.

asaites commented on optipy/src/lib.rs:
```diff
@@ -0,0 +1,223 @@
+extern crate proc_macro;
```
I've updated the `README` with details.

asaites commented on quil-rs/pyproject.toml:
```diff
@@ -22,7 +22,9 @@ dependencies = ["numpy>=1.26"]
-sdist-include = ["README.md"]
+include = [ "README-py.md" ]
+python-source = "python"
+module-name = "quil._quil"
```
It's `maturin`'s recommendations for how to configure a mixed rust/python project (such as those with type stubs): https://www.maturin.rs/project_layout.html

asaites commented on quil-rs/src/instruction/quilpy.rs:
```diff
@@ -0,0 +1,534 @@
+        n_qubits: u64,
+    ) -> PyResult<Bound<'py, PyArray2<Complex64>>> {
+        let py = slf.py();
+        let mut slf = { slf.get().clone() };
```
In short, yes, but I've since refactored it.

asaites commented on quil-rs/python/quil/waveforms.pyi:
```diff
@@ -0,0 +1,453 @@
+        r"""
+        Explicit detuning to bake into iq values
+        """
+    def __new__(cls, duration:builtins.float, fwhm:builtins.float, t0:builtins.float, anh:builtins.float, alpha:builtins.float, sample_rate:builtins.float, second_order_hrm_coeff:builtins.float, scale:builtins.float, phase:builtins.float, detuning:builtins.float) -> HermiteGaussian: ...
```
No, but we probably need to `from __future__ import annotations` at the top of these files. I've run into a few other issues with how `mypy` processes some of these. In many cases, the stubs themselves are valid, but long-standing bugs in `mypy` reject them despite that. I've opened some issues on the stub-gen repo asking if they might make changes for it, but if not, we may need some other handling here.

asaites commented on scripts/findpyclasses.py:
I've since updated the script to avoid the empty output.

asaites commented on quil-rs/src/program/calibration.rs:
```diff
@@ -85,6 +157,10 @@ pub struct CalibrationExpansion {
 impl CalibrationExpansion {
+    pub fn calibration_used(&self) -> &CalibrationSource {
+        &self.calibration_used
+    }
```
Unintentional. I probably originally thought it was not needed (because for Python support, I can just add `#[pyo3(get)]`), then I realized that wasn't true, and added it back in the wrong place. I'll move it back to avoid the weird diff.

asaites commented on quil-rs/src/waveform/templates.rs:
```diff
@@ -35,29 +49,46 @@ fn ceiling_with_epsilon(value: f64) -> f64 {
 }
 
 /// Convert polar coordinates to rectangular coordinates.
+#[allow(dead_code)]
```
I think it's related to [this old issue](https://github.com/rigetti/quil-rs/issues/365), but honestly I don't really know.

kalzoo reviewed (commented):
what a big change - good work.

There is the oddity of python and rust sharing docs, but I'm not too worried about that and I'd like to get this into users' hands asap.

On that note - let's ask for some live validation of this by @bramathon @erichulburd et al before we ship

asaites commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,131 @@
+    steps:
+      - run: echo "Release tag starts with quil-rs/v, proceeding with release"
+  should-publish-wheels:
+    if: (github.event_name == 'workflow_dispatch' && inputs.publishWheels) || (github.event_name == 'release' && !github.event.release.prerelease)
```
OK, this was more of a pain than I expected, but by renaming it back to `publish-quil-rs.yml`, I am able to run it manually. Incidentally, it revealed an issue in how knope updates `quil-cli/Cargo.toml`.

The [`quil-rs` crate published](https://crates.io/crates/quil-rs/0.33.0-rc.2) successfully, but the non-sdist wheels didn't build right, so I will need to dig a moment. I suspect it's due to removing a linker flag (based on some other things I was looking at, I thought we didn't need it, but I am probably mistaken).

asaites commented on scripts/findpyclasses.py:
> Is this the most maintainable approach - a regex parser for Rust written in Python?

That's definitely a fair concern. This script started small and grew quite a lot. I still think the overall approach is reasonable -- munging a directory of text files is very much Python's wheelhouse -- but the regex aspect specifically became more of a pain over time. I think it would be more accurate to use something like `syn` to walk the AST, but I'm not sure that'd necessarily be more maintainable. I think this extraction method was the right "first pass" at it.

> I wonder if something could generate the data we want closer to the truth - such as a wrapper macro around the subject macros that just registers the call

That's actually sorta how `pyo3_stub_gen` works! It might be possible to even use the information it knows to replace what this script is doing, but I'm not certain it has enough information. Also, it obviously wouldn't work anywhere we forgot to add the macro (which is one of the things this script catches).

> Is this under test anywhere?

There is a very small test of a specific part (to verify `Item` comparisons work as intended), but nothing comprehensive. I can add some to verify/demonstrate it catches issues it's intended to.

asaites replied:
I've split the script into a package and a separate script, and subdivided the package into a few key modules. This should make it easier to grok and generally more maintainable, but still it might be sensible to replace the regex with some real tokenization.

I also added a series of high-level functional tests to confirm it catches expected issues. The actual tests are generated based on the contents of the `pyo3_linter/tests/data` directory -- each subdirectory becomes a test, where the code processes the directory contents and compares the discovered annotations/exports/issues to expected ones, stored in a files within that particular test. Those comparison for "issues" is sensitive to the specific text the issue reports, but if that becomes a problem in the future, we can give them actual enumerated values to compare. That seems pretty low priority, though, as it's currently easy to "fix" the tests by just manually confirming the output, then rewriting the "expected-" output file.

asaites commented:
It looks like the build is failing due to linker failures. I have a couple guesses as to why that is, but I'll need to look into it.

asaites commented:
> It looks like the build is failing due to linker failures. I have a couple guesses as to why that is, but I'll need to look into it.

Fixed -- it was because I added a feature flag for the `pyo3/extension-module` (required for proper linking when we build wheels), which causes linking failures otherwise. Since we run a number of tests with `--all-features`, it lead to the CI issues. Putting the feature in the `pyproject.toml` (in the `maturin` config section) should make use of it when we build the wheels, though.

Shadow53 commented on quil-rs/README.md:
```diff
@@ -1,27 +1,198 @@
+    This should reduce many edge cases around subclassing `quil` types,
+    but may constitute a breaking change if you were relying on `__setstate__` support.
+    To be clear, we do not recommend using `pickle` as a serialization format,
+    but understand it has its value for its relationship `copy` and `multiprocessing`.
```
```suggestion
    but understand it has its value for its relationship to `copy` and `multiprocessing`.
```

Shadow53 commented on quil-rs/src/expression/quilpy.rs:
```diff
@@ -0,0 +1,227 @@
+    #[new]
+    fn __new__(value: isize) -> PyResult<Self> {
+        match value {
+            val if val == Self::Caret as isize => Ok(Self::Caret),
```
Could potentially leverage https://crates.io/crates/num_enum for this

Shadow53 reviewed (commented):
Reviewed non-Rust code and skimmed some of the Rust. Will come back to this tomorrow to review the Rust more closely.

asaites commented on quil-rs/Makefile.toml:
```diff
@@ -0,0 +1,75 @@
+[tasks.default]
+alias = "test-all"
+
+[tasks.check-api]
+description = "Check if quil API has breaking changes."
+dependencies = ["install-quil-with-stubs"]
+cwd = ".."
+script = { file = "./scripts/check-quil-api.sh" }
```
My strong preference is to have this task in this `Makefile`, as it is really only a concern for `quil-rs`. I agree that the `cwd` is weird, but its because `griffe` won't operate right unless run from the project root 🥲 . There's a note about this within the script, but perhaps there should be a comment here, too.

I should just move the `scripts` directory into the `quil-rs` for the same reason -- they're only related to `quil-rs`. That won't directly solve the issue (`griffe` still wants to run at the root), but it might make it more obvious the `cwd` is a necessity, despite the script only being relevant to `quil-rs`.

Shadow53 replied:
If you move the `scripts` folder into `quil-rs`, you can also add a `cd` to the script itself instead of needing it in the `Makefile.toml`

asaites replied:
Ok, I've moved the scripts directory and update this script and the Makefile.

Shadow53 reviewed (commented):
"Finished" my previous review. I skipped `pyo3_linter` since it is not published code, so I'm less concerned about it being pristine. Everything else has been reviewed.

asaites commented on .github/workflows/bench.yml:
```diff
@@ -8,52 +8,60 @@ name: Benchmark
-        os: [ubuntu-latest, windows-latest, macOS-latest]
-        rust:
-          - stable
+        platform:
```
I was following a combination of [GitHub's docs on `matrix`](https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/run-job-variations#about-matrix-strategies) and looking at examples of the build processes in other `maturin`-based CI builds, so the short of it is, this is likely just some leftover copypasta.

asaites commented on .github/workflows/publish-quil-rs.yml:
```diff
@@ -1,28 +1,141 @@
+      matrix:
+        python-version: ['3.10', '3.11', '3.12', '3.13']
+        platform:
+          [
+            { os: "ubuntu-latest",  target: "x86_64",  venv: "./quil-rs/.venv", name: "linux" },
+            { os: "ubuntu-latest",  target: "aarch64", venv: "./quil-rs/.venv", name: "linux" },
+            { os: "ubuntu-latest",  target: "ppc64le", venv: "./quil-rs/.venv", name: "linux" },
+            { os: "macos-latest",   target: "universal2-apple-darwin", venv: "./quil-rs/.venv", name: "macos" },
+            { os: "windows-latest", target: "x86_64-pc-windows-msvc",  venv: '.\quil-rs\.venv', name: "windows" },
+          ]
```
In this case, I was basing things off of some other examples, and I preferred this syntax, as it made it more clear to me what was changing between matrix elements.

I can/will convert it to the YAML syntax if you prefer, but I think this is more readable in this case.

Shadow53 replied:
Valid. It just reads odd to me to mix syntax like this, but there's nothing _wrong_ with it, and it makes sense that this makes it easier to see what the difference is between them

asaites commented on .github/workflows/release.yml:
```diff
@@ -11,35 +11,43 @@ on:
+    - uses: actions/checkout@v5
       with:
         fetch-depth: 0
-        token: ${{ secrets.PAT }}
```
It works without it 🤷 . Note that the `env` sets `GITHUB_TOKEN` above.

asaites commented on CONTRIBUTING.md:
```diff
@@ -3,65 +3,146 @@
+To build and test all parts of `quil-rs` and `quil-py`,
+you'll need a [Rust tool chain][rust-tooling] and [Python 3.10+][python-tooling].
+Many of this repository's development tasks have been simplified using [`cargo-make`][],
+which you can install via `Cargo`:
```
I agree -- though I think I'd like to go the other way and remove the monospace so I'm just referring to Cargo-the-tool.

asaites commented on CONTRIBUTING.md:
```diff
@@ -3,65 +3,146 @@
+
+You can use these additional `cargo-make` tasks:
+
+| Task | Description |
+| ---- | ----------- |
+| `generate-stubs` | Generate Python type stubs for `quil` |
```
I don't know about enforcing it, but it's not much trouble for me to update manually.

asaites commented on README.md:
```diff
@@ -1,19 +1,27 @@
 # Quil Parser & Program Builder
 
-This library is the implementation of the [Quil spec](https://github.com/quil-lang/quil) in Rust.
+This repository is the home of [`quil-rs`][] and [`quil-py`][],
```
I'm not sure about releases, but if we do decide to, I think that deserves a separate PR.

asaites commented on quil-rs/Cargo.toml:
```diff
@@ -41,10 +51,15 @@ rand = "0.8.5"
 graphviz-dot = ["dot-writer"]
 wasm-bindgen = []
+python = [ "numpy", "pyo3", "pyo3-build-config" ]
+stubs = [ "python", "pyo3-stub-gen" ]
```
I am resistant to making this particular change for these reasons:

1. I will push back on the idea that it's not clear or obviously related to `python`: all the documentation related to it mentions Python and type stub files, and all the code related to it is necessarily gating a `pyo3-stub-gen` element.
2. The feature gates within the code add a fair bit of line noise, and a longer feature name makes that worse, particularly if `rustfmt` knocks down an attribute into multiple lines.
3. Changing it is not entirely mechanical: even if rust-analyzer/another IDE feature had support for it, there's also the existing documentation to update. Granted, it's not a _huge_ amount of work, but I'm arguing it's not even worth that cost.

That all said, I'm not really that attached to any particular name, so if you feel strongly that it really should be different (vs this being a nit suggestion), I respect that, and will make the change.

asaites commented on quil-rs/Cargo.toml:
```diff
@@ -41,10 +51,15 @@ rand = "0.8.5"
 [features]
 graphviz-dot = ["dot-writer"]
 wasm-bindgen = []
+python = [ "numpy", "pyo3", "pyo3-build-config" ]
+stubs = [ "python", "pyo3-stub-gen" ]
```
I don't think it really makes a difference, but I do like the clarity. I'll do the same a couple lines up for `dot-writer`.

asaites commented on quil-rs/Makefile.toml:
```diff
@@ -0,0 +1,75 @@
+description = "Install the python package with up-to-date stubs (not needed for testing)."
+dependencies = ["generate-stubs", "install-quil"]
+
+[tasks.pytest]
+description = "Run python tests."
+dependencies = ["install-quil"]
+command = "poetry"
+args = ["run", "pytest"]
```
No, the tests are actually under the `tests_py` directory (I think they'd otherwise get included in the Python package, but don't quote me on that), which `pytest` will appropriate pick up from here. Running from the `python` directory would need to specify `pytest ../` to discover them. Note that the task depends on `install-quil`, which is mainly to ensure you've built the latest version, but has the side effect of also "testing" that importing `quil` works as expected.

asaites commented on .github/workflows/publish-quil-rs.yml:
```diff
@@ -1,28 +1,141 @@
+      - uses: actions/checkout@v5
+      - uses: Swatinem/rust-cache@v2
+      - run: |
+          cargo publish -p quil-rs --locked --token ${{ secrets.CRATES_IO_TOKEN }}
```
Are you OK with pushing this off as a separate issue for now?

asaites commented on quil-rs/README.md:
```diff
@@ -1,27 +1,198 @@
+The code related to the Python package is sometimes referred to as `quil-py`,
+both for historical reasons, and to distinguish it from other parts of `quil-rs`.
+
+Note that `maturin` commands (`maturin develop`, `maturin release`, etc.)
```
I don't know if it's new, but [`maturin` lets you configure some Cargo options](https://www.maturin.rs/config.html#cargo-options) within the `pyproject.toml`, which I think is a sensible choice here. I think technically we need the config here anyway for source installs to work correctly, but don't quote me on that.

As a hedge, I set the features manually in CI when building.

I'll also say, I experimented with adding our own `extension-module` feature to just enable `pyo3/extension-module` (based on [this FAQ entry](https://pyo3.rs/v0.26.0/faq.html#i-cant-run-cargo-test-or-i-cant-build-in-a-cargo-workspace-im-having-linker-issues-like-symbol-not-found-or-undefined-reference-to-_pyexc_systemerror) in PyO3's documentation), but that breaks `cargo hack --feature-powerset`, whereas putting it in the `pyproject.toml` bypasses the necessity for additional args to tell `cargo-hack` about it.

asaites commented on quil-rs/README.md:
```diff
@@ -1,27 +1,198 @@
+    (because a Python object's hash is required to be stable throughout its lifetime),
+    but now it's enforced by marking the `#[pyclass]` as `frozen`.
+    If you previously were mutating these types, it was a bug, unfortunately.
+    Nevertheless, removing `__hash__` constitutes a breaking change.
```
I need to reword this a bit. This is really two different breaking change, done for the same reason/constitute the same bug: some types were both mutable and defined `__hash__`, but [das ist verboten](https://docs.python.org/3/reference/datamodel.html#object.__hash__). Most have become immutable (via `frozen`) while others have removed `__hash__`, and both are breaking changes.

asaites commented on quil-rs/python/quil/__init__.pyi:
```diff
@@ -0,0 +1,33 @@
+    r"""
+    Errors which can occur when converting a Quil item to a string.
+    """
+    ...
```
> Seems like the former should just be a ValueError

This is poorly documented, and I need to improve that. The `QuilValueError` is used specifically to report an inappropriate `int` given to `__new__` of a simple enumeration class. Though we _could_ reuse `ValueError`, this is a bit more domain-specific.

Moreover, in my experience with Python, it's quite valuable to have all errors a package raises derive from a single root error defined with the package, even when there exists a builtin exception type that serves a similar purpose. In particular, the actual exception type really only matters if the user intends to catch it and do something with it, so using builtin exceptions risks catching errors unintentionally. I'd argue there's little reason to catch either `QuilValueError` or an ordinary `ValueError` -- **except** when testing and asserting that a particular thing triggers an excepted error. In that case I _really_ do not want to accidentally catch an unrelated builtin exception and mistakenly pass my test. Knowing that can't happen is valuable enough that I'd go so far as to say it's a code smell for a package to raise a builtin exception directly. That said, I think it's reasonable to raise a custom error derived from a more-specific builtin error than `Exception`, and if I had my druthers, I'd write

```python
class QuilValueError(QuilError, ValueError):
    """Raised when an argument to a quil function has an inappropriate value."""
```

And then someone can write catch `QuilValueError`, `QuilError`, or `ValueError` depending on their needs. Unfortunately, due to CPython implementation details, the documentation strongly discourages subclassing more than one exception type, so I do not usually do that. Regardless, I'm am strongly pro-custom exception hierarchy within a package.

> the latter sounds like an interface to convert into a QuilError rather than an error from converting to a Quil string. Maybe ToQuilStringError

I can make that happen.

asaites commented on quil-rs/src/lib.rs:
```diff
@@ -48,3 +48,142 @@ pub mod validation;
+/// which includes the `#[pymethods]` block with its `#[new]` and `__getnewargs__` methods,
+/// and a second that for when the `python` feature is not enabled,
+/// which simply implements the constructor.
+macro_rules! pickleable_new {
```
It's often used to share the same body for the Rust and Python constructors while ensuring `__getnewargs__` will be compatible with the Python constructor. If the`python` feature is not enabled, it simply outputs the Rust constructor.

I think it'd be sensible for this to become a derive macro, too, but it'd need some care to allow the right customization. Definitely that's not something I'm in a hurry for.

antalsz replied:
Nightly rust has support for `macro_rules!` attribute (rust-lang/rust#143547) and derive (rust-lang/rust#143549) macros – once those are stable they’d be an easy win here, since they could be crate-local

asaites commented on quil-rs/src/program/mod.rs:
```diff
@@ -86,65 +94,48 @@ type Result<T> = std::result::Result<T, ProgramError>;
 
-    /// Like `Clone`, but does not clone the body instructions.
+    /// Return a deep copy of the `Program`, but without the body instructions.
     pub fn clone_without_body_instructions(&self) -> Self {
```
I'd like to leave this name alone for now, since it'd be a breaking change for both `quil-rs` and `quil` otherwise, and I expect in the near future we'll see other breaking changes to improve `Program`'s API, likely to include a different approach to instruction manipulation.

asaites commented on quil-rs/src/program/quilpy.rs:
```diff
@@ -0,0 +1,877 @@
+    m.add_class::<PyProgramCalibrationExpansionSourceMap>()?;
+    m.add_class::<PyProgramCalibrationExpansionSourceMapEntry>()?;
+    m.add_class::<PyScheduleSeconds>()?;
+    m.add_class::<ScheduleSecondsItem>()?;
```
That's what `pyo3_linter` is for. I need to add a `Makefile.toml` task for it, but you can do:

```
python3 ../scripts/findpyclasses.py --show-mistakes
```

If I comment out one of these lines, running that gives this:

```
  Couldn't find export for 'quil.program.TimeSpanSeconds'
     (struct 'TimeSpanSeconds' in src/program/quilpy.rs@760)

 1 potential issue(s) discovered.
```

It catches several other issues, but works textually and isn't perfect. In particular, it has to be taught `macro_rules!`, since the only other choice is operating on the expanded version, which is partially out of our control.

asaites commented on quil-rs/src/program/quilpy.rs:
```diff
@@ -0,0 +1,877 @@
+        self.instructions.to_vec()
+    }
+
+    /// Return a deep copy of the `Program`.
+    fn copy(&self) -> Self {
+        self.clone()
+    }
```
No, `Program` is already `deepcopy`able.

asaites commented on quil-rs/src/quilpy/mod.rs:
```diff
@@ -0,0 +1,193 @@
+/// - PyO3's Complex Enums: https://pyo3.rs/v0.25.1/class#complex-enums
+/// - Issue regarding `__qualname__`: https://github.com/PyO3/pyo3/issues/5270
+/// - Python's `inspect`: https://docs.python.org/3/library/inspect.html#inspect.getmembers
+pub(crate) fn fix_enum_qual_names<'py>(typ: &Bound<'py, PyType>) -> PyResult<()> {
```
My expectation is this will be faster and safer, but really I'm just hoping they make that change upstream.

asaites commented on quil-rs/src/waveform/quilpy.rs:
```diff
@@ -0,0 +1,214 @@
+    /// This raises a `ValueError` if `sample_count` is zero.
+    #[new]
+    pub fn __new__(phase: f64, scale: f64, sample_count: u64) -> PyResult<Self> {
+        if sample_count == 0 {
+            return Err(PyValueError::new_err("sample_count must be positive"));
+        }
```
I'm writing this based on what I remember (but didn't look up again): I'm pretty sure this came from the `quil-py` side, but wasn't ever in the `quil-rs` side. In general `quil-rs` doesn't do this kind of validation. I'm not sure about `NonZeroU64`, but I'll give that a shot and see -- if not, I think this should use a custom error type for the reasons I outlined before (or maybe not do validation here?).

asaites commented on quil-rs/python/quil/__init__.pyi:
```diff
@@ -0,0 +1,33 @@
+from . import program
+from . import waveforms
+
+class PickleError(builtins.QuilError):
```
I made this change, then updated the tests and realized that this name makes more sense: you'd most likely use it like `try ... except quil.PickleError:`, or write `from quil import PickleError`. Prefixing it with `Quil` just seems like an unneeded stutter. For the same reason, `QuilValueError` really should just be `ValueError`, and most often used as `quil.ValueError` (if for some reason you needed to explicitly catch it, which is unlikely).
